### ISSUE TITLE
Bug: [no-unsafe-return] any to any callback is not allowed when cb is union type

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#8982]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwxilQGdEcYBbAQVQE8BhAIwAowmAueFl4KDKF2J0AlPAC8APnjCRIofQDcAWABQhYmQo16zHnwELRRidIDea%2BPDgZkMfAaiKrAehfwA8gGk1AXxEqqmqgkLAIKOjYeAREpORUtIxMHjAAcsgQEGyc8KgZEPAAPty8-IIy9GJSFaLyNYEacdqJzCnpmfplRnXCpvBm8JbWILb28I7ObvAAqqRQiAg2dvg4iBXwAAbCGwR0AA4gwPAAblAQyCAAdH4BakA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolrRE%2BWNCbpIiaNA7RI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=true

### Repro Code

```TypeScript
declare function transformAnyCb(cb: ((data: any) => any)): any;
transformAnyCb((data: any): any => {
  return data;  // OK
});

declare function transformAnyCbOrNull(cb: null | ((data: any) => any)): any;
transformAnyCbOrNull((data: any): any => { 
  return data; // Unsafe return of an `any` typed value.
});
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-return":"error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I typed the callback to allow any as a return value. So the any to any converter should be fine.

### Actual Result

ts-eslint is unhappy with 

> Unsafe return of an `any` typed value.

**Comment by kirkwaiblinger** [comment]
@HolgerJeromin Thanks for the report! 🙂 Agreed that this is a bug. When `any` is explicitly specified as a return type, the rule should allow `any` to be returned!
